File: WebAdmin/mwhWebAdmin/Configuration/SeoValidationConfig.cs
using System.ComponentModel.DataAnnotations;

namespace mwhWebAdmin.Configuration;

/// <summary>
/// Centralized SEO validation configuration that ensures consistent rules
/// across all validation systems (PowerShell, C#, JavaScript, and LLM prompts)
/// </summary>
public static class SeoValidationConfig
{
    /// <summary>
    /// Title validation rules
    /// </summary>
    public static class Title
    {
        public const int MinLength = 30;
        public const int MaxLength = 60;
        public const int OptimalMinLength = 40;
        public const int OptimalMaxLength = 55;
        public const int ScoreWeight = 2;

        public static string GetValidationMessage(int length)
        {
            if (length < MinLength)
                return $"Title is too short ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            if (length > MaxLength)
                return $"Title is too long ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            return string.Empty;
        }

        public static int GetScore(int length)
        {
            if (length < MinLength || length > MaxLength)
                return length < MinLength ? 60 : 70;
            return 100;
        }
    }

    /// <summary>
    /// Meta description validation rules
    /// </summary>
    public static class MetaDescription
    {
        public const int MinLength = 150;
        public const int MaxLength = 320;
        public const int OptimalMinLength = 150;
        public const int OptimalMaxLength = 160;
        public const int ScoreWeight = 2;

        public static string GetValidationMessage(int length)
        {
            if (length < MinLength)
                return $"Description is too short ({length} chars). Recommended: {MinLength}-{OptimalMaxLength} characters";
            if (length > MaxLength)
                return $"Description is too long ({length} chars). Recommended: {MinLength}-{OptimalMaxLength} characters (max {MaxLength})";
            return string.Empty;
        }

        public static int GetScore(int length)
        {
            if (length < MinLength || length > MaxLength)
                return length < MinLength ? 60 : 70;
            return 100;
        }
    }

    /// <summary>
    /// Keywords validation rules
    /// </summary>
    public static class Keywords
    {
        public const int MinCount = 3;
        public const int MaxCount = 8;
        public const int OptimalMinCount = 4;
        public const int OptimalMaxCount = 6;
        public const int ScoreWeight = 1;

        public static string GetValidationMessage(int count)
        {
            if (count < MinCount)
                return $"Consider adding more keywords. Current: {count}, Recommended: {MinCount}-{MaxCount}";
            if (count > MaxCount)
                return $"Too many keywords may dilute SEO value. Current: {count}, Recommended: {MinCount}-{MaxCount}";
            return string.Empty;
        }

        public static int GetScore(int count)
        {
            if (count == 0)
                return 80; // Keywords are recommended but not required
            if (count < MinCount)
                return 70;
            if (count > MaxCount)
                return 80;
            return 100;
        }
    }

    /// <summary>
    /// Open Graph description validation rules
    /// Based on research: Official Open Graph Protocol specifies "one to two sentence description"
    /// with no mandatory character limits; 100-300 range provides optimal balance
    /// </summary>
    public static class OpenGraphDescription
    {
        public const int MinLength = 100;
        public const int MaxLength = 300;
        public const int OptimalMinLength = 100;
        public const int OptimalMaxLength = 200;
        public const int ScoreWeight = 1;

        public static string GetValidationMessage(int length)
        {
            if (length < MinLength)
                return $"Open Graph description is too short ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            if (length > MaxLength)
                return $"Open Graph description is too long ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            return string.Empty;
        }

        public static int GetScore(int length)
        {
            if (length < MinLength || length > MaxLength)
                return length < MinLength ? 60 : 70;
            return 100;
        }
    }

    /// <summary>
    /// Twitter description validation rules
    /// Based on research: Twitter's official maximum is 200 characters;
    /// 120-200 range accommodates both concise and detailed descriptions
    /// </summary>
    public static class TwitterDescription
    {
        public const int MinLength = 120;
        public const int MaxLength = 200;
        public const int OptimalMinLength = 120;
        public const int OptimalMaxLength = 180;
        public const int ScoreWeight = 1;

        public static string GetValidationMessage(int length)
        {
            if (length < MinLength)
                return $"Twitter description is too short ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            if (length > MaxLength)
                return $"Twitter description is too long ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            return string.Empty;
        }

        public static int GetScore(int length)
        {
            if (length < MinLength || length > MaxLength)
                return length < MinLength ? 60 : 70;
            return 100;
        }
    }

    /// <summary>
    /// Open Graph title validation rules
    /// </summary>
    public static class OpenGraphTitle
    {
        public const int MinLength = 30;
        public const int MaxLength = 65;
        public const int OptimalMinLength = 30;
        public const int OptimalMaxLength = 60;
        public const int ScoreWeight = 1;

        public static string GetValidationMessage(int length)
        {
            if (length < MinLength)
                return $"Open Graph title is too short ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            if (length > MaxLength)
                return $"Open Graph title is too long ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            return string.Empty;
        }

        public static int GetScore(int length)
        {
            if (length < MinLength || length > MaxLength)
                return length < MinLength ? 60 : 70;
            return 100;
        }
    }

    /// <summary>
    /// Twitter title validation rules
    /// </summary>
    public static class TwitterTitle
    {
        public const int MinLength = 10;
        public const int MaxLength = 50;
        public const int OptimalMinLength = 20;
        public const int OptimalMaxLength = 45;
        public const int ScoreWeight = 1;

        public static string GetValidationMessage(int length)
        {
            if (length < MinLength)
                return $"Twitter title is too short ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            if (length > MaxLength)
                return $"Twitter title is too long ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            return string.Empty;
        }

        public static int GetScore(int length)
        {
            if (length < MinLength || length > MaxLength)
                return length < MinLength ? 60 : 70;
            return 100;
        }
    }

    /// <summary>
    /// H1 tag validation rules
    /// </summary>
    public static class H1Tag
    {
        public const int MinLength = 10;
        public const int MaxLength = 70;
        public const int OptimalMinLength = 20;
        public const int OptimalMaxLength = 60;
        public const int ScoreWeight = 1;

        public static string GetValidationMessage(int length)
        {
            if (length < MinLength)
                return $"H1 text is too short ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            if (length > MaxLength)
                return $"H1 text is too long ({length} chars). Recommended: {MinLength}-{MaxLength} characters";
            return string.Empty;
        }

        public static int GetScore(int length)
        {
            if (length < MinLength || length > MaxLength)
                return length < MinLength ? 60 : 70;
            return 100;
        }
    }

    /// <summary>
    /// Image validation rules
    /// </summary>
    public static class Images
    {
        public const int ScoreWeight = 1;

        public static int GetScore(int totalImages, int imagesWithoutAlt)
        {
            if (totalImages == 0)
                return 100;

            if (imagesWithoutAlt > 0)
            {
                var percentage = (imagesWithoutAlt * 100) / totalImages;
                return Math.Max(100 - (percentage * 2), 20);
            }

            return 100;
        }
    }

    /// <summary>
    /// Overall scoring configuration
    /// </summary>
    public static class Scoring
    {
        /// <summary>
        /// Calculates weighted overall score
        /// Title, Description, and HTML SEO are most important (weighted 2x)
        /// Others are standard weight (1x)
        /// </summary>
        public static int CalculateOverallScore(
            int titleScore,
            int descriptionScore,
            int keywordsScore,
            int imageScore,
            int h1Score,
            int contentImageScore,
            int htmlSeoScore)
        {
            var weightedScore = (titleScore * Title.ScoreWeight +
                               descriptionScore * MetaDescription.ScoreWeight +
                               keywordsScore * Keywords.ScoreWeight +
                               imageScore * Images.ScoreWeight +
                               h1Score * H1Tag.ScoreWeight +
                               contentImageScore * Images.ScoreWeight +
                               htmlSeoScore * 2) / 9.0;

            return Math.Min((int)Math.Round(weightedScore, MidpointRounding.AwayFromZero), 100);
        }
    }

    /// <summary>
    /// Grade calculation configuration
    /// </summary>
    public static class Grading
    {
        public const int GradeAThreshold = 90;
        public const int GradeBThreshold = 80;
        public const int GradeCThreshold = 70;
        public const int GradeDThreshold = 60;

        /// <summary>
        /// Gets the grade based on warnings and overall score
        /// Grade A: 90+ score with NO warnings about length issues
        /// Grade B: 80+ score or 90+ with warnings
        /// Grade C: 70+ score
        /// Grade D: 60+ score
        /// Grade F: Below 60 score
        /// </summary>
        public static string GetGrade(int overallScore, List<string> warnings)
        {
            // Check for any warnings about too long or too short attributes (excluding social media)
            var hasCoreContentLengthWarnings = warnings.Any(w =>
                (w.Contains("too long") || w.Contains("too short") ||
                 w.Contains("Title is too") || w.Contains("Description is too") ||
                 w.Contains("Meta description too") || w.Contains("HTML title too") ||
                 w.Contains("HTML meta description too")) &&
                !w.Contains("Twitter") && !w.Contains("Open Graph"));

            // Grade A only for scores 90+ with no core content length warnings
            if (overallScore >= GradeAThreshold && !hasCoreContentLengthWarnings)
                return "A";

            if (overallScore >= GradeBThreshold)
                return "B";
            if (overallScore >= GradeCThreshold)
                return "C";
            if (overallScore >= GradeDThreshold)
                return "D";

            return "F";
        }
    }

    /// <summary>
    /// Call-to-action words for meta descriptions
    /// </summary>
    public static class CallToAction
    {
        public static readonly string[] Words =
        {
            "discover", "learn", "explore", "understand", "master", "guide",
            "unlock", "reveal", "uncover", "find", "get", "start", "begin",
            "create", "build", "develop", "improve", "optimize", "enhance"
        };

        public static bool HasCallToAction(string text)
        {
            return Words.Any(word => text.ToLowerInvariant().Contains(word));
        }
    }

    /// <summary>
    /// Validation attribute factory methods for consistent model validation
    /// </summary>
    public static class ValidationAttributes
    {
        public static StringLengthAttribute CreateTitleValidation()
        {
            return new StringLengthAttribute(Title.MaxLength)
            {
                MinimumLength = Title.MinLength,
                ErrorMessage = $"Title must be between {Title.MinLength}-{Title.MaxLength} characters"
            };
        }

        public static StringLengthAttribute CreateMetaDescriptionValidation()
        {
            return new StringLengthAttribute(MetaDescription.MaxLength)
            {
                MinimumLength = MetaDescription.MinLength,
                ErrorMessage = $"Meta description must be between {MetaDescription.MinLength}-{MetaDescription.MaxLength} characters"
            };
        }

        public static StringLengthAttribute CreateOpenGraphTitleValidation()
        {
            return new StringLengthAttribute(OpenGraphTitle.MaxLength)
            {
                MinimumLength = OpenGraphTitle.MinLength,
                ErrorMessage = $"Open Graph title must be between {OpenGraphTitle.MinLength}-{OpenGraphTitle.MaxLength} characters"
            };
        }

        public static StringLengthAttribute CreateOpenGraphDescriptionValidation()
        {
            return new StringLengthAttribute(OpenGraphDescription.MaxLength)
            {
                MinimumLength = OpenGraphDescription.MinLength,
                ErrorMessage = $"Open Graph description must be between {OpenGraphDescription.MinLength}-{OpenGraphDescription.MaxLength} characters"
            };
        }

        public static StringLengthAttribute CreateTwitterTitleValidation()
        {
            return new StringLengthAttribute(TwitterTitle.MaxLength)
            {
                MinimumLength = TwitterTitle.MinLength,
                ErrorMessage = $"Twitter title must be between {TwitterTitle.MinLength}-{TwitterTitle.MaxLength} characters"
            };
        }

        public static StringLengthAttribute CreateTwitterDescriptionValidation()
        {
            return new StringLengthAttribute(TwitterDescription.MaxLength)
            {
                MinimumLength = TwitterDescription.MinLength,
                ErrorMessage = $"Twitter description must be between {TwitterDescription.MinLength}-{TwitterDescription.MaxLength} characters"
            };
        }
    }
}

File: WebAdmin/mwhWebAdmin/wwwroot/js/seo-validation-config.js
/**
 * Centralized SEO validation configuration for JavaScript
 * Keeps frontend validation consistent with backend C# configuration
 */
const SeoValidationConfig = {
    Title: {
        MinLength: 30,
        MaxLength: 60,
        OptimalMinLength: 40,
        OptimalMaxLength: 55,
        ScoreWeight: 2,

        getValidationMessage: function(length) {
            if (length < this.MinLength)
                return `Title is too short (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            if (length > this.MaxLength)
                return `Title is too long (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            return '';
        },

        getScore: function(length) {
            if (length < this.MinLength || length > this.MaxLength)
                return length < this.MinLength ? 60 : 70;
            return 100;
        }
    },

    MetaDescription: {
        MinLength: 120,
        MaxLength: 160,
        OptimalMinLength: 120,
        OptimalMaxLength: 160,
        ScoreWeight: 2,

        getValidationMessage: function(length) {
            if (length < this.MinLength)
                return `Description is too short (${length} chars). Recommended: ${this.MinLength}-${this.OptimalMaxLength} characters`;
            if (length > this.MaxLength)
                return `Description is too long (${length} chars). Recommended: ${this.MinLength}-${this.OptimalMaxLength} characters (max ${this.MaxLength})`;
            return '';
        },

        getScore: function(length) {
            if (length < this.MinLength || length > this.MaxLength)
                return length < this.MinLength ? 60 : 70;
            return 100;
        }
    },

    Keywords: {
        MinCount: 3,
        MaxCount: 8,
        OptimalMinCount: 4,
        OptimalMaxCount: 6,
        ScoreWeight: 1,

        getValidationMessage: function(count) {
            if (count < this.MinCount)
                return `Consider adding more keywords. Current: ${count}, Recommended: ${this.MinCount}-${this.MaxCount}`;
            if (count > this.MaxCount)
                return `Too many keywords may dilute SEO value. Current: ${count}, Recommended: ${this.MinCount}-${this.MaxCount}`;
            return '';
        },

        getScore: function(count) {
            if (count === 0)
                return 80; // Keywords are recommended but not required
            if (count < this.MinCount)
                return 70;
            if (count > this.MaxCount)
                return 80;
            return 100;
        }
    },

    OpenGraphDescription: {
        MinLength: 120,
        MaxLength: 160,
        OptimalMinLength: 120,
        OptimalMaxLength: 160,
        ScoreWeight: 1,

        getValidationMessage: function(length) {
            if (length < this.MinLength)
                return `Open Graph description is too short (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            if (length > this.MaxLength)
                return `Open Graph description is too long (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            return '';
        },

        getScore: function(length) {
            if (length < this.MinLength || length > this.MaxLength)
                return length < this.MinLength ? 60 : 70;
            return 100;
        }
    },

    TwitterDescription: {
        MinLength: 120,
        MaxLength: 160,
        OptimalMinLength: 120,
        OptimalMaxLength: 160,
        ScoreWeight: 1,

        getValidationMessage: function(length) {
            if (length < this.MinLength)
                return `Twitter description is too short (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            if (length > this.MaxLength)
                return `Twitter description is too long (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            return '';
        },

        getScore: function(length) {
            if (length < this.MinLength || length > this.MaxLength)
                return length < this.MinLength ? 60 : 70;
            return 100;
        }
    },

    OpenGraphTitle: {
        MinLength: 30,
        MaxLength: 65,
        OptimalMinLength: 30,
        OptimalMaxLength: 60,
        ScoreWeight: 1,

        getValidationMessage: function(length) {
            if (length < this.MinLength)
                return `Open Graph title is too short (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            if (length > this.MaxLength)
                return `Open Graph title is too long (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            return '';
        },

        getScore: function(length) {
            if (length < this.MinLength || length > this.MaxLength)
                return length < this.MinLength ? 60 : 70;
            return 100;
        }
    },

    TwitterTitle: {
        MinLength: 10,
        MaxLength: 50,
        OptimalMinLength: 20,
        OptimalMaxLength: 45,
        ScoreWeight: 1,

        getValidationMessage: function(length) {
            if (length < this.MinLength)
                return `Twitter title is too short (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            if (length > this.MaxLength)
                return `Twitter title is too long (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            return '';
        },

        getScore: function(length) {
            if (length < this.MinLength || length > this.MaxLength)
                return length < this.MinLength ? 60 : 70;
            return 100;
        }
    },

    H1Tag: {
        MinLength: 10,
        MaxLength: 70,
        OptimalMinLength: 20,
        OptimalMaxLength: 60,
        ScoreWeight: 1,

        getValidationMessage: function(length) {
            if (length < this.MinLength)
                return `H1 text is too short (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            if (length > this.MaxLength)
                return `H1 text is too long (${length} chars). Recommended: ${this.MinLength}-${this.MaxLength} characters`;
            return '';
        },

        getScore: function(length) {
            if (length < this.MinLength || length > this.MaxLength)
                return length < this.MinLength ? 60 : 70;
            return 100;
        }
    },

    Images: {
        ScoreWeight: 1,

        getScore: function(totalImages, imagesWithoutAlt) {
            if (totalImages === 0)
                return 100;

            if (imagesWithoutAlt > 0) {
                const percentage = (imagesWithoutAlt * 100) / totalImages;
                return Math.max(100 - (percentage * 2), 20);
            }

            return 100;
        }
    },

    Scoring: {
        calculateOverallScore: function(titleScore, descriptionScore, keywordsScore, imageScore, h1Score, contentImageScore, htmlSeoScore) {
            const weightedScore = (titleScore * SeoValidationConfig.Title.ScoreWeight +
                                 descriptionScore * SeoValidationConfig.MetaDescription.ScoreWeight +
                                 keywordsScore * SeoValidationConfig.Keywords.ScoreWeight +
                                 imageScore * SeoValidationConfig.Images.ScoreWeight +
                                 h1Score * SeoValidationConfig.H1Tag.ScoreWeight +
                                 contentImageScore * SeoValidationConfig.Images.ScoreWeight +
                                 htmlSeoScore * 2) / 9.0;

            return Math.min(Math.round(weightedScore), 100);
        }
    },

    Grading: {
        GradeAThreshold: 90,
        GradeBThreshold: 80,
        GradeCThreshold: 70,
        GradeDThreshold: 60,

        getGrade: function(overallScore, warnings) {
            // Check for any warnings about too long or too short attributes (excluding social media)
            const hasCoreContentLengthWarnings = warnings.some(w =>
                (w.includes('too long') || w.includes('too short') ||
                 w.includes('Title is too') || w.includes('Description is too') ||
                 w.includes('Meta description too') || w.includes('HTML title too') ||
                 w.includes('HTML meta description too')) &&
                !w.includes('Twitter') && !w.includes('Open Graph'));

            // Grade A only for scores 90+ with no core content length warnings
            if (overallScore >= this.GradeAThreshold && !hasCoreContentLengthWarnings)
                return 'A';

            if (overallScore >= this.GradeBThreshold)
                return 'B';
            if (overallScore >= this.GradeCThreshold)
                return 'C';
            if (overallScore >= this.GradeDThreshold)
                return 'D';

            return 'F';
        }
    },

    CallToAction: {
        Words: [
            'discover', 'learn', 'explore', 'understand', 'master', 'guide',
            'unlock', 'reveal', 'uncover', 'find', 'get', 'start', 'begin',
            'create', 'build', 'develop', 'improve', 'optimize', 'enhance'
        ],

        hasCallToAction: function(text) {
            return this.Words.some(word => text.toLowerCase().includes(word));
        }
    }
};

// Character count update function using centralized config
function updateCharacterCount(elementId, text, configSection) {
    const element = document.getElementById(elementId);
    if (!element) return;

    const length = text.length;
    const config = SeoValidationConfig[configSection];
    if (!config) return;

    let className = 'text-success';

    if (length < config.MinLength || length > config.MaxLength) {
        className = 'text-danger';
    } else if (length < (config.MinLength + 10) || length > (config.MaxLength - 10)) {
        className = 'text-warning';
    }

    element.textContent = `${length} characters (${config.MinLength}-${config.MaxLength} recommended)`;
    element.className = className;
}

// Keyword count update function
function updateKeywordCount(elementId, keywords) {
    const element = document.getElementById(elementId);
    if (!element) return;

    const keywordList = keywords.split(',').map(k => k.trim()).filter(k => k.length > 0);
    const count = keywordList.length;
    const config = SeoValidationConfig.Keywords;

    let className = 'text-success';

    if (count < config.MinCount || count > config.MaxCount) {
        className = 'text-danger';
    } else if (count < config.OptimalMinCount || count > config.OptimalMaxCount) {
        className = 'text-warning';
    }

    element.textContent = `${count} keywords (${config.MinCount}-${config.MaxCount} recommended)`;
    element.className = className;
}

// Export for use in other scripts
if (typeof module !== 'undefined' && module.exports) {
    module.exports = SeoValidationConfig;
}
